### PR TITLE
Custom Door Assets and Dependencies

### PR DESCRIPTION
**Pull request by toasterparty**
Create and adds custom door assets and list their dependencies. This includes the meta enum `DoorType` which is like the door version of `PickupType`.

Nothing added is actually used in this PR. This is just prerequisites for later.

If you would prefer, we can consolidate the resources into a single pool. Plando fork is written with a separate resource pool for doors and there isn't anything too ugly as a consequence.

### CODE REVIEW

File: src/patches.rs [hunk 1]
```diff
@@ -89,25 +95,165 @@ fn collect_pickup_resources<'r>(gc_disc: &structs::GcDisc<'r>, starting_items: &
...
+
+    assert!(looking_for.is_empty());
+
+    found
+}
+
+
+// Door assets are not shared across all areas either,
+// so we have to make a cache for them as well.
+fn collect_door_resources<'r>(gc_disc: &structs::GcDisc<'r>)
```
Comment (aprilwade) [inline src/patches.rs]: Why is this separate from `collect_pickup_resources`/why is `door_resources` seperate from `pickup_resources`?

File: src/patches.rs [hunk 2]
```diff
@@ -89,25 +95,165 @@ fn collect_pickup_resources<'r>(gc_disc: &structs::GcDisc<'r>, starting_items: &
...
+    -> HashMap<(u32, FourCC), structs::Resource<'r>>
+{   
+    // Get list of all dependencies needed by custom doors //
+    
+    let mut looking_for = HashSet::<_>::new();
+
+    {
+        let looking_for_door: HashSet<_> = DoorType::iter()
```
Comment (aprilwade) [inline src/patches.rs]: Can this block be written as `looking_for.extend(DoorType::iter().flat_map(...))` and `looking_for.extend(BlastShieldType::iter().flat_map(...))`?

File: src/patches.rs [hunk 3]
```diff
@@ -89,25 +95,165 @@ fn collect_pickup_resources<'r>(gc_disc: &structs::GcDisc<'r>, starting_items: &
...
+    {
+        println!("error - still looking for {:?}", looking_for);
+    }
+
     assert!(looking_for.is_empty());
 
     found
 }
 
+fn create_custom_door_cmdl<'r>(
```
Comment (aprilwade) [inline src/patches.rs]: This probably belongs in custom_assets.rs along side the functions that make custom CMDLs.

File: src/patches.rs [hunk 4]
```diff
@@ -89,25 +95,165 @@ fn collect_pickup_resources<'r>(gc_disc: &structs::GcDisc<'r>, starting_items: &
...
+    // Remove extra assets from dependency search since they won't appear     //
+    // in any pak. Instead add them to the output resource pool. These assets //
+    // are provided as external files checked into the repository.            //
     for res in crate::custom_assets::custom_assets(&found, starting_items) {
         let key = (res.file_id, res.fourcc());
         looking_for.remove(&key);
         assert!(found.insert(key, res).is_none());
     }
 
+    if !looking_for.is_empty()
```
Comment (aprilwade) [inline src/patches.rs]: Brace should be on the same line as the `if` here.

File: src/patches.rs [hunk 5]
```diff
@@ -89,25 +95,165 @@ fn collect_pickup_resources<'r>(gc_disc: &structs::GcDisc<'r>, starting_items: &
...
+
+    // Add the newly generated resources //
+    for res in new_assets {
+        let key = (res.file_id, res.fourcc());
+        if looking_for.remove(&key) {
+            assert!(found.insert(key, res).is_none());
+        }
+    }
+
+    if !looking_for.is_empty()
```
Comment (aprilwade) [inline src/patches.rs]: Brace should be on the same line as the if here.

File: src/patches.rs (hunk 1 above)
Comment (toasterparty) [reply src/patches.rs]: That was how Yonic did it. All of the plando features I've added either need `door_resources` or `pickup_resources`, but never both. If you would like, I can refactor everything to use a single resource pool.

**Comment by aprilwade** [comment]
Yes, I think a single pool is a better idea. Feel free to rename to something more appropriate than pickup_resources

File: src/patches.rs (hunk 1 above)
Comment (toasterparty) [reply src/patches.rs]: I created `game_resources` as a catch-all

File: src/patches.rs (hunk 2 above)
Comment (toasterparty) [reply src/patches.rs]: I'm not sure how you could specify to map just the dependencies here, and not all of the type enums

**Comment by toasterparty** [comment]
@aprilwade Please review once more. I couldn't not figure out a nicer way to add all of the dependencies to the `looking_for` vec, but that's probably due to lack of experience. Let me know if you come up with something, otherwise, this PR is ready.

**Pull request merged**